dataactbroker/handlers/fileGenerationHandler.py

```diff
@@ -58,57 +77,90 @@ def generate_d_file(file_type, agency_code, start, end, job_id, file_name, uploa
             upload_name - Filename to use on S3
             is_local - True if in local development, False otherwise
     """
-    logger.debug('Starting file {} generation'.format(file_type))
-
-    with job_context(job_id) as session:
-        file_utils = fileD1 if file_type == 'D1' else fileD2
-        full_file_path = "".join([CONFIG_BROKER['d_file_storage_path'], file_name])
-        page_idx = 0
+    with job_context(job_id, is_local) as sess:
+        # check if FileRequest already exists with this job_id
+        current_date = datetime.now().date()
+        file_request = sess.query(FileRequest).filter_by(job_id=job_id).one_or_none()
+        if not file_request:
+            file_request = FileRequest(request_date=current_date, job_id=job_id, start_date=start, end_date=end,
+                                       agency_code=agency_code, file_type=file_type, is_cached_file=False)
+            sess.add(file_request)
+        elif file_request.request_date != current_date:
+            # current file is cached, but not from today
+            # un-cache it and update its date
+            file_request.request_date = current_date
+            file_request.is_cached_file = False
+
+        # search for separate FileRequest to mark as parent
+        parent_req = None
+        if not file_request.is_cached_file:
+            parent_req = sess.query(FileRequest).\
+                filter_by(request_date=current_date, file_type=file_type, start_date=start, end_date=end,
+                          agency_code=agency_code, is_cached_file=True).one_or_none()
+            file_request.parent_job_id = parent_req.job_id if parent_req else None
+        sess.commit()
+
+        if file_request.is_cached_file:
+            # this is the cached file, no need to do anything
+            logger.debug('{} file has already been generated by this job'.format(file_type))
+        elif parent_req:
+            # copy parent data to this job if parent is not still running
+            if parent_req.job.job_status_id != JOB_STATUS_DICT['running']:
+                copy_parent_file_request_data(sess, file_request.job, parent_req.job, file_type, is_local)
+        else:
+            # no cached file
+            # start D file generation
+            logger.debug('Starting file {} generation'.format(file_type))
+            file_utils = fileD1 if file_type == 'D1' else fileD2
+            full_file_path = "".join([CONFIG_BROKER['d_file_storage_path'], file_name])
+            page_idx = 0
 
-        try:
```

Comment:
Why did we remove the try/finally?

Reply:
Including that `try/finally` caused a bug that I found. If it threw an error during that period, it wouldn't reach the error handling in `job_context()`, which means the job wouldn't ever fail correctly and it would run forever.